mirrord-layer/src/file/filter.rs

```diff
@@ -0,0 +1,304 @@
+
+/// Holds the `Regex` that is used to either continue or bypass file path operations (such as
+/// `file::ops::open`), according to what the user specified.
+///
+/// The `FileFilter::Include` variant takes precedence and erases whatever the user supplied as
+/// exclude, this means that if the user specifies both, `FileFilter::Exclude` is never constructed.
+///
+/// Warning: Use `FileFilter::new` (or equivalent) when initializing this, otherwise the above
+/// constraint might not be held.
+#[derive(Debug, Clone)]
+pub(crate) enum FileFilter {
+    /// User specified `Regex` containing the file paths that the user wants to include for file
+    /// operations.
+    ///
+    /// Overrides `FileFilter::Exclude`.
+    Include(Regex),
+
+    /// Combination of `DEFAULT_EXCLUDE_LIST` and the user's specified `Regex`.
+    ///
+    /// Anything not matched by this `Regex` is considered as included.
+    Exclude(Regex),
+}
+
+impl FileFilter {
+    /// Initializes a `FileFilter` based on the user configuration.
+    ///
+    /// - `FileFilter::Include` is returned if the user specified any include path (thus erasing
+    ///   anything passed as exclude);
+    /// - `FileFilter::Exclude` also appends the `DEFAULT_EXCLUDE_LIST` to the user supplied regex;
+    #[tracing::instrument(level = "debug")]
+    pub(crate) fn new(user_config: FileFilterConfig) -> Self {
+        let FileFilterConfig { include, exclude } = user_config;
+        let include = include.map(VecOrSingle::to_vec).unwrap_or_default();
+        let exclude = exclude.map(VecOrSingle::to_vec).unwrap_or_default();
+
+        let default_exclude = DEFAULT_EXCLUDE_LIST.as_ref();
+        let default_exclude_regex =
+            Regex::new(default_exclude).expect("Failed parsing default exclude file regex!");
+
+        // Converts a list of `String` into one big regex-fied `String`.
+        let reduce_to_string = |list: Vec<String>| {
+            list.into_iter()
+                // Turn into capture group `(/folder/first.txt)`.
+                .map(|element| format!("({element})"))
+                // Put `or` operation between groups `(/folder/first.txt)|(/folder/second.txt)`.
+                .reduce(|acc, element| format!("{acc}|{element}"))
+        };
+
+        let exclude = reduce_to_string(exclude)
+            // Add default exclude list.
+            .map(|mut user_exclude| {
+                user_exclude.push_str(&format!("|{default_exclude}"));
+                user_exclude
+            })
+            .as_deref()
+            .map(Regex::new)
+            .transpose()
+            .expect("Failed parsing exclude file regex!")
+            .unwrap_or(default_exclude_regex);
+
+        reduce_to_string(include)
+            .as_deref()
+            .map(Regex::new)
+            .transpose()
+            .expect("Failed parsing include file regex!")
+            .map(Self::Include)
+            .unwrap_or(Self::Exclude(exclude))
```

Comment:
When would unwrap fail here?

Reply:
And/or in the command before it

Reply:
The above part: if the user specifies no `exclude`, then we `unwrap` as the `default_exclude_regex`.

Below: when the user specifies no `include`, we `unwrap` as the `exclude` (which is guaranteed to exist either as a user specified, or a default value).